Feat/UI overhaul

Fenriir42 opened the pull request.

Fenriir42 committed:
docs: add IRON-VOX UI design specification

Add comprehensive design guide with:
- Tekkadan (dark) and Barbatos (light) theme palettes
- Color system specifications and usage guidelines
- UI component specifications (armor cards, buttons, sliders)
- Layout blueprints for settings, library, and karaoke modes
- Implementation tips for egui/Rust

Fenriir42 committed:
feat(ui): implement Tekkadan and Barbatos theme system

Add dual-theme color system:
- Tekkadan (dark mode): Void green background, flower red primary
- Barbatos (light mode): Hangar wall background, cobalt blue primary
- 8 color methods: background, card_surface, primary, secondary, accent, alert, text_primary, text_muted
- Theme toggle functionality
- Apply method to set egui context visuals

Fenriir42 committed:
feat(ui): implement top navigation and bottom playback panels

Add top panel:
- IRON-VOX branding
- Navigation tabs (Library/Karaoke/Settings)
- Theme switcher button

Add bottom panel:
- Album art placeholder
- Song info display
- Playback controls (play/pause, skip forward/backward)
- Piston-style progress bar with handle
- Duration display (current/total)
- Volume control placeholder

Fenriir42 committed:
feat(ui): add armor card widget and library components

Add armor card widget:
- Chamfered container with 3px left border accent
- 16px padding for content spacing

Add library section components:
- Header card with album art placeholder and track info
- Path input with manual entry and scan button
- Song list with clickable items and hover effects
- Track number, name, duration, and mic icon display
- Filtered search support

Fenriir42 committed:
feat(ui): implement settings panel with four sections

Add settings panel structure:
- Sidebar with 4 categories (Audio/Display/Library/Network)
- Audio section: device selection, latency, mic calibration
- Input gain slider with percentage display
- Mechanical switch-style noise gate toggle
- Theme override radio buttons
- Settings cards with headers and separators
- Factory reset and save config actions

All sections use stubbed functionality for future implementation

Fenriir42 committed:
feat(app): implement three-view navigation system

Add AppView enum for Library/Karaoke/Settings views
Update app state with:
- SettingsState for settings panel
- library_path_input for manual path entry (WSL workaround)

Add view-specific render methods:
- render_library_view with sidebar and main content
- render_karaoke_view (stubbed)
- render_settings_view

Update main loop to handle:
- Theme switching
- View changes
- Playback actions (play/pause/stop/skip)
- Library scanning with file dialog fallback

Apply responsive layouts with margins and proper sizing

Fenriir42 committed:
feat(audio): implement proper pause/resume with timing tracking

Add timing state tracking:
- start_time for playback start
- pause_time for pause moment
- accumulated_time for total played time

Update methods:
- pause() saves accumulated time and freezes position
- resume() restarts timer from accumulated time
- get_position() returns correct time when paused or playing
- start_tracking() and reset_position() for proper state management

This fixes the pause button functionality and progress bar timing

Fenriir42 committed:
feat(ui): add optional custom font support for CaskaydiaMono

Add custom-font feature flag to Cargo.toml
Add setup_fonts() method with conditional compilation:
- When custom-font enabled: loads CaskaydiaMono Nerd Font
- Otherwise: uses default egui fonts with UTF-8 support

Add FONT_SETUP.md with installation instructions
Create assets directory for font file placement

This enables better UTF-8 character display and Nerd Font icons support

Fenriir42 committed:
refactor: remove redundant comments across codebase

Remove unnecessary comments that don't provide crucial information:
- File header comments
- Obvious section markers
- Self-explanatory code annotations
- Redundant inline explanations

Keep code self-documenting through clear naming and structure
All remaining code maintains readability without verbose comments

Fenriir42 committed:
fix: address clippy warnings and linting issues

- Add allow(dead_code) for Song::extension field (future use)
- Add allow(dead_code) for PlaybackAction variants (stubbed)
- Change filter parameter from &mut String to &mut str for better API

All clippy checks now pass with -D warnings

Fenriir42 committed:
chore: merge dev into feat/ui_overhaul to sync with latest changes

Fenriir42 committed:
fix: address clippy warnings after merge

Ciznia commented on docs/Design_Guide.md:
```diff
@@ -0,0 +1,193 @@
+Instead of a raw military simulation, imagine this is a **civilian music player built with military-grade parts**.
+
+* **Legibility First:** Decorative elements are background texture only. Text must pop.
+* **Familiar Layouts:** We adhere to standard mental models.
```
Have you done some research about standard mental models and some benchmark about if we respect it

Ciznia commented on src/app.rs:
```diff
@@ -4,29 +4,36 @@ use tracing::{error, info, warn};
 use crate::library::{scanner, Song};
-use crate::ui::{panels, theme::Theme, widgets};
+use crate::ui::{panels, settings::SettingsState, theme::Theme, widgets};
 
 pub struct KaraokeApp {
-    // UI State
-    counter: i32,
-    user_text: String,
     theme: Theme,
+    current_view: AppView,
+    settings_state: SettingsState,
 
-    // Audio
     audio_player: AudioPlayer,
     is_playing: bool,
     current_file: Option<PathBuf>,
     error_message: Option<String>,
     song_duration: Option<std::time::Duration>,
 
-    // Library
     library: Vec<Song>,
     library_path: Option<PathBuf>,
     library_filter: String,
+    library_path_input: String,
+}
```
Maybe do smaller structure to avoid a "trash structure" and split this into smaller structure to have only related properties

Ciznia commented on src/app.rs:
```diff
@@ -47,21 +54,43 @@ impl KaraokeApp {
+        }
+    }
+
+    #[allow(dead_code)]
```
why this pragma

Fenriir42 replied:
Because if not then it clippy will drop kick me about unused bit of code, but they're supposed to be there for now so that i don't have to rewrite 48515849 things in the ui when a feature is finished to link it

Ciznia reviewed (changes requested):
Of what i understand, there is a lot dead_code juste remove it, and some button have color constrast problem (at least in dark mode)

Ciznia committed:
refactor: split into smaller structure to avoid a `trash` struct

Fenriir42 committed:
fix(audio): ensure playback state is updated correctly when audio player is empty

Fenriir42 committed:
feat(theme): integrate enum_cycling for theme cycling functionality

Fenriir42 committed:
feat: add installer metadata and Music category for cargo-bundle

Fenriir42 committed:
feat: add multi-platform installer build system

- Add build script for Linux/macOS/Windows installers
- Support for .deb, AppImage, .dmg, and .msi formats
- Include Windows cross-compilation from Linux
- Add icon assets and documentation
- Comprehensive README with prerequisites and instructions

Fenriir42 committed:
feat: add installer build targets to Makefile

- Add installer-deps target with OpenSSL detection
- Add installer target for building packages
- Include helpful error messages for missing dependencies

Fenriir42 committed:
docs: add installation and build instructions for installers

Fenriir42 committed:
ci: add GitHub Actions workflow for multi-platform releases

- Build installers for Linux, Windows, and macOS
- Support manual workflow dispatch for testing
- Automatic release creation on version tags
- Upload all installers as artifacts

Fenriir42 committed:
feat: add dependencies for persistent library storage

- Add chrono for timestamp tracking
- Add dirs for cross-platform app data directories
- Update Cargo.lock with new dependencies

Fenriir42 committed:
feat: implement persistent library storage system

- Add storage module for managing song library files
- Store songs in platform-specific app data directories
- Track metadata (original path, added date, title) in JSON
- Copy imported songs with unique filenames to prevent conflicts
- Support loading and saving library state across sessions

Fenriir42 committed:
feat: integrate persistent library with UI

- Load library automatically on startup from storage
- Add songs via file dialog or path input
- Remove songs from library storage
- Update library metadata after add/remove operations
- Auto-rescan library directory after changes

Fenriir42 committed:
feat: update library UI for persistent storage

- Replace folder scanning with persistent library management
- Add path input field for adding songs (WSL-friendly)
- Add remove button for each song in library
- Add error color to theme for delete actions
- Remove unused ScanFolder/ScanFromInput actions

Fenriir42 committed:
chore: add dev_library to gitignore

Fenriir42 committed:
style: apply formatting fixes

- Add consistent line breaks in markdown lists
- Format Rust code with rustfmt
- No functional changes

Ciznia commented on src/ui/theme.rs:
```diff
@@ -1,79 +1,113 @@
-    pub fn toggle(self) -> Self {
+impl EnumCycle for Theme {
+    fn up(&self) -> Self {
         match self {
-            Theme::IronFlower => Theme::WhiteFrame,
-            Theme::WhiteFrame => Theme::IronFlower,
+            Theme::Tekkadan => Theme::Barbatos,
+            Theme::Barbatos => Theme::Tekkadan,
         }
     }
 
+    fn down(&self) -> Self {
+        self.up()
+    }
+}
```
Not sure that how the crate for enum cycle work since you still manually switch

Fenriir42 replied:
yes it's how it's supposed to work because the crate only provides the trait, not a derive macro, so i have to implement up & down
